// common/save.js
var { deleteLocalDocument } = require('common/document-helpers')
var template = require('./save.html')

module.exports = function({ stateRouter, mediator, documentName, documentStateName, name, route, messageName }) {
	function cleanUpDocumentAndGoToMainMenu() {
		mediator.publish('finishDocument', documentName, () => {
			deleteLocalDocument(documentName)
			stateRouter.go('scanner')
		})
	}

	stateRouter.addState({
		name,
		route,
		template: {
			template,
			twoway: false
		},
		resolve: function(data, parameters, cb) {
			mediator.request('fetchDocument', documentName, (err, doc) => {
				if (err) {
					cb(err)
				} else if (!doc) {
					cb.redirect(documentStateName)
				} else {
					cb(err, {
						doc: doc.store.getState(),
						back: documentStateName,
						documentName: documentName
					})
				}
			})
		},
		activate: function({ domApi }) {
			mediator.publish('emitToServer', messageName, domApi.get('doc'), err => {
				if (err) {
					domApi.set('error', err)
					console.error(err)
				} else {
					cleanUpDocumentAndGoToMainMenu()
				}
			})
		}
	})
}
